File: internal/schemas/schemas.go
package schemas

// AgentManifest defines a monitoring agent's metadata and context.
type AgentManifest struct {
	Name          string   `yaml:"name"`
	DisplayName   string   `yaml:"display_name"`
	Description   string   `yaml:"description"`
	TargetOS      string   `yaml:"target_os"`
	Ports         []int    `yaml:"ports"`
	Services      []string `yaml:"services"`
	Prerequisites []string `yaml:"prerequisites"`
}

// DiagnosticHints holds agent-specific diagnostic commands and context.
type DiagnosticHints struct {
	PriorityCommands []string `yaml:"priority_commands"`
	ContextHints     []string `yaml:"context_hints"`
}

// DiagnosticPayload is Turn 1 output: LLM tells us what diagnostic commands to run.
type DiagnosticPayload struct {
	Hypothesis         string   `json:"hypothesis"`
	DiagnosticCommands []string `json:"diagnostic_commands"`
}

// RemediationPayload is Turn 2 output: LLM provides the fix after seeing diagnostic results.
type RemediationPayload struct {
	RootCause          string `json:"root_cause"`
	HumanExplanation   string `json:"human_explanation"`
	RemediationCommand string `json:"remediation_command"`
	IsDestructive      bool   `json:"is_destructive"`
}

// StepResult captures the result of executing a single script step.
type StepResult struct {
	StepNumber int    `json:"step_number"`
	Command    string `json:"command"`
	ExitCode   int    `json:"exit_code"`
	Stdout     string `json:"stdout"`
	Stderr     string `json:"stderr"`
	Success    bool   `json:"success"`
}

// RunbookEntry represents a single resolved issue in the runbook.
type RunbookEntry struct {
	ID            string `yaml:"id"`
	ErrorPattern  string `yaml:"error_pattern"`
	StepFailed    string `yaml:"step_failed"`
	RootCause     string `yaml:"root_cause"`
	FixCommand    string `yaml:"fix_command"`
	ResolvedCount int    `yaml:"resolved_count"`
	FirstSeen     string `yaml:"first_seen"`
	LastSeen      string `yaml:"last_seen"`
}

// RunbookIndex is the lookup table mapping error patterns to entry files.
type RunbookIndex struct {
	Entries []RunbookIndexEntry `yaml:"entries"`
}

// RunbookIndexEntry maps an error pattern to a runbook entry file.
type RunbookIndexEntry struct {
	Pattern   string `yaml:"pattern"`
	EntryFile string `yaml:"entry_file"`
}

File: nr_diagnose/schemas.py
"""Data schemas for the diagnostic system."""

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class RequiredInput:
    """A user-supplied value needed by the install script (e.g. credentials, hostnames)."""
    name: str = ""
    description: str = ""
    secret: bool = False
    default: str = ""


@dataclass
class AgentManifest:
    name: str = ""
    display_name: str = ""
    description: str = ""
    target_os: str = ""
    ports: List[int] = field(default_factory=list)
    services: List[str] = field(default_factory=list)
    prerequisites: List[str] = field(default_factory=list)
    required_inputs: List[RequiredInput] = field(default_factory=list)


@dataclass
class DiagnosticHints:
    priority_commands: List[str] = field(default_factory=list)
    context_hints: List[str] = field(default_factory=list)


@dataclass
class DiagnosticPayload:
    """Turn 1 output: LLM tells us what diagnostic commands to run."""
    hypothesis: str = ""
    diagnostic_commands: List[str] = field(default_factory=list)


@dataclass
class RemediationPayload:
    """Turn 2 output: LLM provides the fix after seeing diagnostic results."""
    root_cause: str = ""
    human_explanation: str = ""
    remediation_command: str = ""
    is_destructive: bool = False
    # Names of user-supplied inputs (from manifest required_inputs) that caused this failure.
    # When set, the runner offers an [u]pdate option that re-prompts and re-renders affected steps.
    bad_inputs: List[str] = field(default_factory=list)


@dataclass
class StepResult:
    """Captures the result of executing a single script step."""
    step_number: int = 0
    command: str = ""
    exit_code: int = 0
    stdout: str = ""
    stderr: str = ""
    success: bool = False


@dataclass
class RunbookEntry:
    """A single resolved issue in the runbook."""
    id: str = ""
    error_pattern: str = ""
    step_failed: str = ""
    root_cause: str = ""
    fix_command: str = ""
    resolved_count: int = 0
    first_seen: str = ""
    last_seen: str = ""


@dataclass
class RunbookIndexEntry:
    """Maps an error pattern to a runbook entry file."""
    pattern: str = ""
    entry_file: str = ""


@dataclass
class RunbookIndex:
    """Lookup table mapping error patterns to entry files."""
    entries: List[RunbookIndexEntry] = field(default_factory=list)
